In [1]:
!git clone https://github.com/febse/ta2025.git repo

fatal: destination path 'repo' already exists and is not an empty directory.


# Linear Algebra Essentials

Embed a local video:

{{< video videos/VectorMovement.mp4 >}}

## Vectors: Arrows, Lists

Imagine that you have two bank accounts: one in Bulgarian levs (BGN) and one in Euros (EUR). You want to keep track of how much money you have in each account which is to say you want to keep track of two numbers. You start with 0 BGN and 0 EUR and deposit 2 BGN and 1 EUR.

$$
\begin{bmatrix} 0 \quad \text{BGN} \\ 0 \quad \text{EUR} \end{bmatrix} + \begin{bmatrix} 1 \quad \text{BGN} \\ 2 \quad \text{EUR} \end{bmatrix} = \begin{bmatrix} 0 + 1 \quad \text{BGN} \\ 0 + 2  \quad \text{EUR}\end{bmatrix}
$$

You end up with a new list of two numbers: 2 BGN and 1 EUR. To simplify the notation we can drop the currency labels and just write the numbers in a column but keep in mind that the first number corresponds to BGN and the second to EUR. The amounts in both accounts will then look like a table with two rows and one column, let's call it vector and name it $\vec{v}$:

$$
\vec{v} = \begin{bmatrix} 1 \\ 2 \end{bmatrix}
$$

Somewhat unusually for accounting but you can associate this collection of values with a geometric object: an arrow in a plane pointing from the origin (0, 0) to the point (1, 2).


## Sums and Differences of Vectors

Imagine that after the initial deposit you receive a payment of 2 BGN and send a transfer of 1 EUR. You can represent these as movement instructions in the BGN-EUR plane:

- Move 2 units right (increase BGN by 2)
- Move 1 unit down (decrease EUR by 1)

After this the new balances will be 2 BGN and 1 EUR which you can represent or with our list notation:

$$
\begin{bmatrix} 1 \\ 2 \end{bmatrix} + \
\begin{bmatrix} 2 \\ -1 \end{bmatrix} = \begin{bmatrix} 1 + 2 \\ 2 + (-1) \end{bmatrix} = \begin{bmatrix} 3 \\ 1 \end{bmatrix}
$$

This is the same as if you directly write down the _net_ deposits into the accounts: 3 BGN and 1 EUR. In other words, the sum of the two vectors gives you the coordinates of the final point after following the movement instructions so that you can go to it directly (in one step) from the origin.


Differences of vectors work in exactly the same way. Let's view the second bank operation and multiply it by -1. This will mean that instead of receiving 2 BGN and sending 1 EUR you will send 2 BGN and receive 1 EUR so the final balance after that operation will be -1 BGN (you owe 1 BGN to your bank) and 3 EUR:

$$
\begin{bmatrix} 1 \\ 2 \end{bmatrix} - \
\begin{bmatrix} 2 \\ -1 \end{bmatrix} = \begin{bmatrix} 1 - 2 \\ 2 - (-1) \end{bmatrix} = \begin{bmatrix} -1 \\ 3 \end{bmatrix}
$$

## Scalar Multiplication

Let's continue with our bank example and the two accounts. Our first step was to deposit 1 BGN and 2 EUR and we obtained the list of balances

$$
\begin{bmatrix} 1 \\ 2 \end{bmatrix}
$$

Let's simply repeat this operation one more time: we again deposit 1 BGN and 2 EUR. The new balances will be:

$$
\begin{bmatrix} 1 + 1 \\ 2 + 2 \end{bmatrix} = \begin{bmatrix} 2 \cdot 1 \\ 2 \cdot 2 \end{bmatrix} = \begin{bmatrix} 2 \\ 4 \end{bmatrix}
$$

We can write this more concisely as a multiplication of the original vector by the number 2:

$$
2 \cdot \begin{bmatrix} 1 \\ 2 \end{bmatrix} = \begin{bmatrix} 2 \cdot 1 \\ 2 \cdot 2 \end{bmatrix} = \begin{bmatrix} 2 \\ 4 \end{bmatrix}
$$

In general, multiplying a vector by a number (called scalar) means multiplying each entry of the vector by that number. Geometrically this means stretching the arrow by that factor (if the factor is greater than 1) or shrinking it (if the factor is between 0 and 1). If the factor is negative, the arrow is also flipped to point in the opposite direction.


In [3]:
import numpy as np

try:
    from animations.rotation import show_transformation, heart_param_points, heart_line_color
except ImportError:
    from repo.animations.rotation import show_transformation, heart_param_points, heart_line_color


def generate_scaling_matrices(scale_factors):
    """Generate scaling matrices for a sequence of scale factors.
    
    Args:
        scale_factors: array or list of scale factors (uniform scaling)
    
    Returns:
        List of matrix lists, where each element is [S(scale)] for that scale factor
    """
    matrices_sequence = []
    for s in scale_factors:
        # Uniform scaling matrix
        S = np.array([[s, 0],
                      [0, s]])
        matrices_sequence.append([S])
    return matrices_sequence


# Create a range of scale factors from -2.0 to 2.0 (including negative values)
scale_factors = np.linspace(-2.0, 2.0, 81)

# Use heart points as the input set
pts = heart_param_points(n=60)

# Generate scaling matrices
matrices_sequence = generate_scaling_matrices(scale_factors)

# Show the transformation animation
fig = show_transformation(
    matrices_sequence=matrices_sequence,
    points=pts,
    L=5,
    title='Scaling Transformation: Interactive Slider',
    color_func=heart_line_color,
    frame_duration_ms=100
)

# Update slider labels to show scale factors instead of rotation angles
slider_steps = []
for idx, s in enumerate(scale_factors):
    step = {
        'label': f'{s:.2f}',
        'method': 'animate',
        'args': [[f'frame_{idx}'], {
            'mode': 'immediate',
            'frame': {'duration': 0, 'redraw': True},
            'transition': {'duration': 0}
        }]
    }
    slider_steps.append(step)

# Update the slider configuration
fig.update_layout(
    sliders=[{
        'active': 0,
        'currentvalue': {'prefix': 'Scale Factor: ', 'suffix': '', 'visible': True},
        'pad': {'t': 40},
        'steps': slider_steps
    }]
)

fig.show()

## Vector Magnitude

While the direction of a vector is an instruction of where to go, the magnitude (length) of the vector tells us how far to go in that direction.
We can compute it by using the Pythagorean theorem (extended to multiple dimensions). With two dimensions we have for the $v = (1, 2)^T$ vector:

$$
||v|| = \sqrt{1^2 + 2^2} = \sqrt{5} \approx 2.236
$$



## Vector Similarity


We consider two vectors to be similar if they point in similar directions. One way to measure this is to compute the cosine of the angle between them. This is called cosine similarity. The cosine theorem is an extension of the Pythagorean theorem and states that in a triangle with sides of lengths a, b and c and angle γ opposite to side c it holds that:

$$
c^2 = a^2 + b^2 - 2ab\cos(\theta)
$$

It can be extended to show the relation between the cosine between two vectors and their dot product.

$$
\cos(\theta) = \frac{u \cdot v}{||u|| \, ||v||}
$$

The cosine of any angle is restricted between -1 and 1 so you can think about it as the percentage of similarity between the directions in which the two vectors point. If the cosine is 1, the vectors point in the same direction, if it is -1 they point in opposite directions and if it is 0 they are orthogonal (perpendicular).

The dot product of two vectors is simply the sum of the products of their corresponding entries (so it is only defined for vectors of the same length).

For example:

$$
u = \begin{pmatrix} 1 \\ 2 \\ 3 \end{pmatrix}, \quad v = \begin{pmatrix} 1 \\ -1 \\ 0 \end{pmatrix} \quad \Rightarrow \quad u \cdot v = 1 \cdot 1 + 2 \cdot (-1) + 3 \cdot 0 = -1
$$

By the way, we can also view the dot product as the matrix multiplication of a 1 x n matrix (row vector) with an n x 1 matrix (column vector).

$$
u^T v = \begin{pmatrix} 1 & 2 & 3 \end{pmatrix} \begin{pmatrix} 1 \\ -1 \\ 0 \end{pmatrix} = -1
$$

To find the cosine of the two vectors we need their magnitudes as well:

$$
||u|| = \sqrt{1^2 + 2^2 + 3^2} = \sqrt{14} \approx 3.742
$$

$$
||v|| = \sqrt{1^2 + (-1)^2 + 0^2} = \sqrt{2} \approx 1.414
$$

Now we are ready to compute the cosine:

$$
\cos(\theta) = \frac{-1}{3.742 \cdot 1.414} \approx -0.189
$$

In [2]:
# The same with numpy

import numpy as np

u = np.array([1, 2, 3])
v = np.array([1, -1, 0])

dot_product = np.dot(u, v)
print(f"Dot product of u and v using np.dot: {dot_product}")

Dot product of u and v using np.dot: -1


In [3]:
# The lengths (magnitudes) are

print(f"Length of u: {np.linalg.norm(u).round(3)}")
print(f"Length of v: {np.linalg.norm(v).round(3)}")

Length of u: 3.742
Length of v: 1.414


In [4]:
# So the cosine similarity is

cosine_similarity = dot_product / (np.linalg.norm(u) * np.linalg.norm(v))
print(f"Cosine similarity between u and v: {cosine_similarity.round(3)}")

Cosine similarity between u and v: -0.189


In [5]:
# Is it the same as the correlation?

np.corrcoef(u, v)

array([[ 1. , -0.5],
       [-0.5,  1. ]])

The correlation between two vectors is actually defined quite similarly.

$$
\rho_{u,v} = \frac{\sum_{i=1}^{n} (u_i - \bar{u})(v_i - \bar{v})}{\sqrt{\sum_{i=1}^{n} (u_i - \bar{u})^2} \sqrt{\sum_{i=1}^{n} (v_i - \bar{v})^2}}
$$

Both the correlation and the cosine similarity are restricted between -1 and 1 (i.e you can think about them as percentages). What is the difference then?

## Linear Transformations

A lot of real-life operations involve simple transformations such as for example calculating the average or the total balance of multiple bank accounts in different currencies.

Let's view the total balance of two bank accounts in BGN and EUR (in BGN) after the initial deposits (1 BGN and 2 EUR). To get the total balance in BGN we need to bring the EUR amount to BGN by multiplying it by the exchange rate (1 EUR = 1.95583 BGN) and then summing the two amounts.

$$
\text{Total in BGN} = \underset{\text{BGN/BGN}}{1} \cdot \underset{\text{amt BGN}}{1} + \underset{\text{BGN/EUR}}{1.95583} \cdot \underset{\text{amt EUR}}{2} = 4.91166 \, \text{BGN}
$$

Let's generalize this operation a bit. Let's generalize this operation a bit
by leaving the exact values of the amounts unspecified and writing them as variables $x$ and $y$. Then a function taking concrete amounts and returning the total in BGN can be written as:

$$
f(x, y) = 1 \cdot x + 1.95583 \cdot y
$$

In [6]:
def f(x, y):
    return 1 * x + 1.95583 * y

# Example usage
f(1, 2)

4.9116599999999995

This function has a couple of properties that make it quite simple and _easily predictable_:

- If you double both amounts, the total will double as well:

$$
f(2x, 2y) = 1 \cdot (2x) + 1.95583 \cdot (2y) = 2 \cdot (1 \cdot x + 1.95583 \cdot y) = 2f(x, y)
$$

In [7]:
f(2 * 1, 2 * 2)  # Should be 2 * f(1, 2)

9.823319999999999


- **Additivity** If you split the amounts into two parts and compute the total for each part separately, the overall total will be the sum of the two totals:

$$
\begin{align*}
f(x_1 + x_2, y_1 + y_2) & = 1 \cdot (x_1 + x_2) + 1.95583 \cdot (y_1 + y_2) \\
& = (1 \cdot x_1 + 1.95583 \cdot y_1) + (1 \cdot x_2 + 1.95583 \cdot y_2) \\
& = f(x_1, y_1) + f(x_2, y_2)
\end{align*}
$$

In [8]:
print("f(1, 2) + f(3, 4)", f(1, 2) + f(3, 4)) # Should be f(1 + 3, 2 + 4)
print("f(1 + 3, 2 + 4)", f(1 + 3, 2 + 4)) 

f(1, 2) + f(3, 4) 15.734979999999998
f(1 + 3, 2 + 4) 15.73498


And lastly, if there is no money in either account, the total is zero:

$$
f(0, 0) = 1 \cdot 0 + 1.95583 \cdot 0 = 0
$$

This is by the way not the case for functions like this one:

$$
g(x, y) = 1 \cdot x + 1.95583 \cdot y + 5
$$

$$
g(0, 0) = 1 \cdot 0^2 + 1.95583 \cdot 0 + 5 = 5 \neq 0
$$

By the way, we have already seen forms like the one in $f(x, y)$. When you compare it with the definition of the dot product you will notice that this nothing more than the dot product of a vector of exchange rates and a vector of amounts.

We can think about the function $f(x, y)$ as taking a vector argument and returning a scalar value:

$$
f\left(\begin{bmatrix} x \\ y \end{bmatrix}\right) = \begin{bmatrix} 1 & 1.95583 \end{bmatrix} \cdot \begin{bmatrix} x \\ y \end{bmatrix}
$$

For the sake of intuition we can name the function $f$ for example "total BGN calculator". What if we want to create another function that calculates the average balance in EUR? It should take the same vector argument (the amounts in BGN and EUR) and return a scalar value (the average balance in EUR). We can write it like this:

$$
h\left(\begin{bmatrix} x \\ y \end{bmatrix}\right) = \frac{1}{2} \left( \underset{\text{EUR/BGN}}{0.51129} \cdot \underset{\text{amt BGN}}{x} + \underset{\text{EUR/EUR}}{1} \cdot \underset{\text{amt EUR}}{y} \right)
$$

or more concisely using the dot product notation:

$$
h\left(\begin{bmatrix} x \\ y \end{bmatrix}\right) = \frac{1}{2} \begin{bmatrix} 0.51129 & 1 \end{bmatrix} \cdot \begin{bmatrix} x \\ y \end{bmatrix}
$$

Now, as we are getting more familiar with these functions, we may want to play around and try to combine them so that we have a single function that returns both the total in BGN and the average in EUR. Because it must return two values we can think about it as returning a vector:

$$
F\left(\begin{bmatrix} x \\ y \end{bmatrix}\right) = \begin{bmatrix} f\left(\begin{bmatrix} x \\ y \end{bmatrix}\right) \\ h\left(\begin{bmatrix} x \\ y \end{bmatrix}\right) \end{bmatrix} = \begin{bmatrix} \begin{bmatrix} 1 & 1.95583 \end{bmatrix} \cdot \begin{bmatrix} x \\ y \end{bmatrix} \\ \frac{1}{2} \begin{bmatrix} 0.51129 & 1 \end{bmatrix} \cdot \begin{bmatrix} x \\ y \end{bmatrix} \end{bmatrix}
$$

Now this expression is full of brackets so that it is a bit hard to read. Keeping in mind what operations are being performed we can drop the brackets and write it more concisely as:

$$
F\left(\begin{bmatrix} x \\ y \end{bmatrix}\right) = \begin{bmatrix} 1 & 1.95583 \\ \frac{1}{2} \cdot 0.51129 & \frac{1}{2} \cdot 1 \end{bmatrix} \cdot \begin{bmatrix} x \\ y \end{bmatrix}
$$

Again, remember that we have two mini-functions, each returning a scalar value by calculating a dot product between a set of weights (exchange rates, averaging factor) and the input vector (amounts in BGN and EUR). The function $F$ simply stacks the two scalar values into a list (vector) and returns it.

In particular, note that this new function $F$ also has the same three properties as the original functions $f$ and $h$:

- If you double both amounts, both the total and the average will double as well:

$$
F\left(\begin{bmatrix} 2x \\ 2y \end{bmatrix}\right) = 2 \cdot F\left(\begin{bmatrix} x \\ y \end{bmatrix}\right)
$$
- **Additivity** If you split the amounts into two parts and compute the total and average for each part separately, the overall total and average will be the sum of the two totals and averages:

$$
F\left(\begin{bmatrix} x_1 + x_2 \\ y_1 + y_2 \end{bmatrix}\right) = F\left(\begin{bmatrix} x_1 \\ y_1 \end{bmatrix}\right) + F\left(\begin{bmatrix} x_2 \\ y_2 \end{bmatrix}\right)
$$

- If there is no money in either account, both the total and the average are zero:
$$
F\left(\begin{bmatrix} 0 \\ 0 \end{bmatrix}\right) = \begin{bmatrix} 0 \\ 0 \end{bmatrix}
$$

Also not another thing: this function takes a vector with two entries and returns a vector with two entries. In general this function maps vectors from a 2-dimensional space to vectors in another 2-dimensional space. To make things more compact, we can ignore the measurement units (BGN and EUR) and think about this function as mapping vectors from $\mathbb{R}^2$ to $\mathbb{R}^2$.

We can also imagine that the function $F$ takes the input and _moves_ it to a new location in the 2D space.

You can also imagine other functions just as $F$ that report different summaries of the amounts in the two accounts such as for example only the amount in BGN, only the amount in EUR. As long as these functions have the three properties mentioned above, they can all be calculated as dot products between a set of weights and the input vector and their outputs can be stacked into a vector and returned together.

$$
G\left(\begin{bmatrix} x \\ y \end{bmatrix}\right) = \begin{bmatrix} 1 & 1.95583 \\ \frac{0.51}{2} & \frac{1}{2} \\ 0 & 1 \\ 1 & 0 \\ 0 & 1.95583 \end{bmatrix} \cdot \begin{bmatrix} x \\ y \end{bmatrix} = 
\begin{bmatrix} 1x + 1.95583y \\ \frac{0.51}{2}x + \frac{1}{2}y \\ 0x + 1y \\ 1x + 0y \\ 0x + 1.95583y \end{bmatrix}
$$

For the sake of intuitive understanding, let's name the operations performed by $G$. The first row computes the total in BGN, the second row computes the average in EUR, the third row extracts the amount in EUR and the fourth row extracts the amount in BGN, the fifth one calculates only the amount in BGN
of the Euro account. As the function returns a list of five elements, we can think about it as mapping vectors from $\mathbb{R}^2$ to vectors in $\mathbb{R}^5$.


When you look at both $F$ and $G$ you will notice all we need to know
about these functions is contained in the tables of weights, so we can simplify our notation even further by just writing down the weight matrices. For example, we can write:

$$
\begin{bmatrix} 1 & 1.95583 \\ \frac{1}{2} \cdot 0.51129 & \frac{1}{2} \cdot 1 \end{bmatrix}
$$

and 

$$
\begin{bmatrix} 1 & 1.95583 \\ \frac{0.51}{2} & \frac{1}{2} \\ 0 & 1 \\ 1 & 0 \\ 0 & 1.95583 \end{bmatrix}
$$

Implicitly, we will understand that these matrices represent functions that take vectors from $\mathbb{R}^2$ and return vectors in $\mathbb{R}^2$ and $\mathbb{R}^5$ respectively by calculating the dot products between their rows and the input vector.

We started the discussion with a very simple (though not entirely artificial) example of an accounting operation but we arrived at a quite general function that works not only with the specific amounts we started with (1 BGN and 2 EUR) but with any amounts in the two accounts. Now, going beyond accounting, we can ask: where does this function move all the vectors in the 2D space? In other words, what is the image of the entire space $\mathbb{R}^2$ under the transformation $F$?

For this it is convenient to plot a couple of vectors and look at where they end up after applying the transformation.

In [15]:
def F(x: np.ndarray) -> np.ndarray:
    return np.array([
        1 * x[0] + 1.95583 * x[1], # The first dot product (total in BGN)
        (0.51 * x[0] + 1 * x[1]) / 2 # The second dot product (average in EUR)
    ])

print("F([1, 2])", F(np.array([1, 2])))
print("F([1, 2])", F(np.array([4, 1])))


F([1, 2]) [4.91166 1.255  ]
F([1, 2]) [5.95583 1.52   ]


## Geometric View of Linear Transformations

Let's take $F$ from the example so far and calculate the image of a few vectors under this transformation. To emphasize the geometric aspect, we want to show the _movement_ of the original vectors to their new locations after applying the transformation (hence the animation).

Let's look at a couple of matrices and see what the do.

$$
S = \begin{bmatrix} 1 & 2 \\ 0 & 1 \end{bmatrix} \quad \text{shear in x-direction}
$$

$$
R = \begin{bmatrix} 0 & -1 \\ 1 & 0 \end{bmatrix} \quad \text{90 degree rotation}
$$

In [7]:
import numpy as np

try:
    from animations.rotation import show_transformation, generate_rotation_matrices, heart_param_points, heart_line_color
except ImportError:
    from repo.animations.rotation import show_transformation, generate_rotation_matrices, heart_param_points, heart_line_color


# Define target transformation matrix M
# Example: a shear + scale
M = np.array([[0, -1],
              [1, 0]])

# Build an interpolation from Identity to M: T(t) = I + t (M - I)
I = np.eye(2)
steps = 25
matrices_sequence = []
for k in range(steps + 1):
    t = k / steps
    T_t = I + t * (M - I)
    matrices_sequence.append([T_t])  # each frame applies one matrix

# Use heart points as the input set
pts = heart_param_points(n=60)

# Show the transformation animation
fig = show_transformation(
    matrices_sequence=matrices_sequence,
    points=pts,
    L=5,
    title='Heart Points: Identity → Target Matrix',
    color_func=heart_line_color,
    frame_duration_ms=300
)
fig.show()

Pay special attention to the locations where the function maps the two vectors

$$
e_1 = \begin{bmatrix} 1 \\ 0 \end{bmatrix}, \quad 
e_2 = \begin{bmatrix} 0 \\ 1 \end{bmatrix}
$$

In [16]:
print("F([1, 0])", F(np.array([1, 0])))
print("F([0, 1])", F(np.array([0, 1])))

F([1, 0]) [1.    0.255]
F([0, 1]) [1.95583 0.5    ]


These vectors are especially convenient because we can easily express any other vector in $\mathbb{R}^2$ by scaling and adding them together.

For example our account balances at the start were:

$$
\begin{bmatrix} 1 \\ 2 \end{bmatrix} = 1 \underset{e_1}{\begin{bmatrix} 1 \\ 0 \end{bmatrix}}
+ 2 \underset{e_2}{\begin{bmatrix} 0 \\ 1 \end{bmatrix}} = 
\begin{bmatrix} 1 \\ 0 \end{bmatrix} + \begin{bmatrix} 0 \\ 2 \end{bmatrix}
= \begin{bmatrix} 1 + 0 \\ 0 + 2 \end{bmatrix}
= \begin{bmatrix} 1 \\ 2 \end{bmatrix}
$$

Here you can also see why the three properties (additivity, homogeneity, and a zero fixed point) are important.

To find the image of any vector under $F$ we only need to track where it 
sends $e_1$ and $e_2$. To see this, look at the example.

$F$ moves $\begin{bmatrix} 1 \\ 2 \end{bmatrix}$ to $\begin{bmatrix} 4.9 \\ 1.2 \end{bmatrix}$. But we can also find this by looking at the transformed basis vectors.

$$
\begin{align*}
e^{*}_1 & = F(e_1) = \begin{bmatrix} 1 \\ 0.25 \end{bmatrix} \\
e^{*}_2 & = F(e_2) = \begin{bmatrix} 1.955 \\ 0.5 \end{bmatrix}
\end{align*}
$$

From the additivity property of $F$ follows that

$$
\begin{align*}
F(v) & = F(1 e_1 + 2 e_2) \\
    &  = F(1e_1) + F(2e_2) \quad \text{additivity} \\
    & = 1 F(e_1) + 2F(e_2) \quad \text{homogeneity} \\
    & = 1 e^{*}_1 + 2 e^{*}_2 
\end{align*}
$$

So the transformed vector can be found as a linear combination of the transformed basis vectors. As the transformed basis vectors are simply the columns of the transformation matrix, we can find the transformed vector by directly weighing and summing the columns of the matrix.

$$
F(v) = 1 \begin{bmatrix} 1 \\ 0.25 \end{bmatrix} + 2 \begin{bmatrix} 1.955 \\ 0.5 \end{bmatrix} = \begin{bmatrix} 4.911 \\ 1.25 \end{bmatrix}
$$

In a more general setting, we can define the transformation of any vector $v$ in $\mathbb{R}^2$ by a matrix $A$ as follows:

$$
A = \begin{bmatrix} a & b \\ c & d \end{bmatrix}, \quad v = \begin{bmatrix} x \\ y \end{bmatrix}
$$

$$
A(v) = x \begin{bmatrix} a \\ c \end{bmatrix} + y \begin{bmatrix} b \\ d \end{bmatrix} = \begin{bmatrix} ax + by \\ cx + dy \end{bmatrix}
$$

You will recognize this as the standard definition of matrix multiplication and in fact, the transformation
is often written simply as:

$$
A v 
$$

because it shares some properties with multiplication of numbers:



- If you double the vector, the result will double as well:

$$
A (2v) = 2 A(v)
$$

- The transformation is associative 

$$
A (B C) v  = (A B) C v
$$

- Distributive over vector addition:

$$
A (u + v) = A(u) + A(v)
$$

It is not commutative (meaning you can't swap the order of matrix multiplications). Generally:

$$
A B v \neq B A v
$$


In [6]:
# Use numpy to 

x = np.array([1, 2])
M = np.array([[1.95583, 1.70], [0, 1]])

M.dot(x)
M @ x

array([5.35583, 2.     ])

In [ ]:
try:
    from animations.rotation import show_rotation
except ImportError:
    from repo.animations.rotation import show_rotation

fig = show_rotation()
fig.show()

## Change of Basis

